src/services/database.ts

```diff
@@ -68,6 +68,89 @@ export class DatabaseService {
     return await db.bookmarks.where('unread').equals(1).toArray();
   }
 
+  static async getBookmarksPaginated(filter: 'all' | 'unread' | 'archived', page: number, pageSize: number): Promise<LocalBookmark[]> {
+    const offset = (page - 1) * pageSize;
+    
+    if (filter === 'unread') {
+      return await db.bookmarks
+        .where('unread').equals(1)
+        .and(bookmark => !bookmark.is_archived)
+        .offset(offset)
+        .limit(pageSize)
+        .toArray();
+    } else if (filter === 'archived') {
+      return await db.bookmarks
+        .where('is_archived').equals(1)
+        .offset(offset)
+        .limit(pageSize)
+        .toArray();
+    } else {
+      // 'all' filter shows only unarchived bookmarks
+      return await db.bookmarks
+        .orderBy('date_added')
+        .reverse()
+        .filter(bookmark => !bookmark.is_archived)
+        .offset(offset)
+        .limit(pageSize)
+        .toArray();
+    }
+  }
+
+  static async getBookmarkCount(filter: 'all' | 'unread' | 'archived'): Promise<number> {
+    if (filter === 'unread') {
+      return await db.bookmarks.where('unread').equals(1).and(bookmark => !bookmark.is_archived).count();
+    } else if (filter === 'archived') {
+      return await db.bookmarks.where('is_archived').equals(1).count();
+    } else {
+      // 'all' filter shows only unarchived bookmarks
+      return await db.bookmarks.filter(bookmark => !bookmark.is_archived).count();
```

Comment:
Refactor the code so that `getBookmarksPaginated()`, `getBookmarkCount()`, `findBookmarkPage()`, etc. all call a shared function to get the filter/where expressions for the filter. That way we won't accidentally end up with these two code paths diverging.